Prose/ProseLib/Sources/ConversationFeature/MessageBar/MessageBar.swift

```diff
@@ -126,22 +158,31 @@ public let messageBarReducer: Reducer<
 // MARK: State
 
 public struct MessageBarState: Equatable {
+  let chatId: JID
+  let loggedInUserJID: JID
+  var typing: [Name]
   var textField: MessageBarTextFieldState
-  var typing: [TypingUser]
   var reactionPicker: ReactionPickerState?
 
   public init(
-    textField: MessageBarTextFieldState,
-    typing: [TypingUser] = []
+    chatId: JID,
+    loggedInUserJID: JID,
+    chatName: Name,
+    message: String = "",
+    typing: [Name] = []
   ) {
-    self.textField = textField
+    self.chatId = chatId
+    self.loggedInUserJID = loggedInUserJID
     self.typing = typing
+    self.textField = .init(chatId: chatId, chatName: chatName, message: message)
   }
 }
 
 // MARK: Actions
 
 public enum MessageBarAction: Equatable, BindableAction {
+  case onAppear, onDisappear
+  case typing([Name])
```

Comment:
`.typing` is incredibly terse. If you derive the name from what actually happens that would be `.participantChatStatesChanged` or after mapping and filtering `.composingUsersChanged`.